File: C#/SensitiveDataTest.cs
using System.Text.RegularExpressions;

namespace SecurityTester.Tests;

public class SensitiveDataTest
{
    private record SensitivePattern(string Name, string Pattern);

    private static readonly List<SensitivePattern> Patterns = new()
    {
        new("Plain-text Password",      @"[""']?password[""']?\s*[:=]\s*[""'][^""']{3,}[""']"),
        new("Credit Card Number",       @"\b(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13})\b"),
        new("Stack Trace / File Path",  @"at\s+\w+.*\.(cs|js):\d+|/home/|/var/www/|C:\\Users\\"),
        new("Database Connection String",@"mongodb://|mysql://|postgres://|Server=.*;Database="),
        new("Internal IP Address",      @"\b(192\.168\.|10\.|172\.(1[6-9]|2\d|3[01])\.)\d+\.\d+\b"),
        new("JWT Token in Response",    @"eyJ[a-zA-Z0-9_-]+\.eyJ[a-zA-Z0-9_-]+\.[a-zA-Z0-9_-]+"),
    };

    private static readonly string[] CookieFlags = { "httponly", "secure" };

    public static async Task<TestResult> RunAsync(string target, HttpClient client, SiteMap? siteMap = null)
    {
        var result  = new TestResult { Name = "Sensitive Data Exposure" };
        var baseUrl = target.TrimEnd('/');
        // Use crawled pages if available, otherwise fall back to common paths
        var pages = siteMap?.Pages.Select(p => p.Url).ToArray()
                    ?? new[] { baseUrl, $"{baseUrl}/api", $"{baseUrl}/api/users", $"{baseUrl}/login" };

        foreach (var page in pages)
        {
            try
            {
                    var res  = await client.GetAsync(page);
                    var body = await res.Content.ReadAsStringAsync();
                    await Task.Delay(100);
                    if ((int)res.StatusCode == 404) continue;

                foreach (var p in Patterns)
                {
                    var match = Regex.Match(body, p.Pattern, RegexOptions.IgnoreCase);
                    if (!match.Success) continue;

                    var raw    = match.Value;
                    var masked = raw.Length > 10 ? raw[..6] + "..." + raw[^4..] : "****";

                    result.Passed = false;
                    result.Issues.Add(new SecurityIssue
                    {
                        Title    = $"Sensitive Data Exposed: {p.Name}",
                        Location = page,
                        Detail   = $"Pattern matched in response: \"{masked}\"",
                        Severity = p.Name.Contains("Password") || p.Name.Contains("Card") ? "CRITICAL" : "HIGH",
                        Risk     = $"Exposing {p.Name.ToLower()} gives attackers direct access to credentials or system internals.",
                        Fixes    = new()
                        {
                            new() { Label = "C# — never return sensitive fields in API responses",
                                    Code  = "// ❌ VULNERABLE — returns everything including password hash\nreturn Ok(user);\n\n// ✅ SAFE — use a DTO (Data Transfer Object)\npublic record UserDto(int Id, string Name, string Email);\nreturn Ok(new UserDto(user.Id, user.Name, user.Email));" },
                            new() { Label = "Use [JsonIgnore] on sensitive model properties",
                                    Code  = "public class User\n{\n    public int    Id    { get; set; }\n    public string Name  { get; set; } = \"\";\n    public string Email { get; set; } = \"\";\n\n    [JsonIgnore]  // Never serialized\n    public string PasswordHash { get; set; } = \"\";\n\n    [JsonIgnore]\n    public string ResetToken { get; set; } = \"\";\n}" },
                            new() { Label = "Never expose stack traces in production",
                                    Code  = "// In Program.cs:\nif (!app.Environment.IsDevelopment())\n{\n    app.UseExceptionHandler(\"/error\");\n}\n\n// Error controller:\n[Route(\"error\")]\npublic IActionResult Error() => Problem(\"An error occurred.\");" }
                        }
                    });
                }

                // Check cookie flags
                if (res.Headers.TryGetValues("Set-Cookie", out var cookies))
                {
                    foreach (var cookie in cookies)
                    {
                        var cl = cookie.ToLower();
                        foreach (var flag in CookieFlags)
                        {
                            if (!cl.Contains(flag))
                            {
                                result.Passed = false;
                                result.Issues.Add(new SecurityIssue
                                {
                                    Title    = $"Cookie Missing \"{flag}\" Flag",
                                    Location = $"Set-Cookie header at {page}",
                                    Detail   = $"Cookie set without \"{flag}\": {cookie.Split(';')[0]}",
                                    Severity = "MEDIUM",
                                    Risk     = flag == "httponly"
                                        ? "JavaScript can read the cookie — XSS can steal session tokens."
                                        : "Cookie sent over plain HTTP — can be intercepted.",
                                    Fixes    = new()
                                    {
                                        new() { Label = "C# — set all security flags on cookies",
                                                Code  = "builder.Services.ConfigureApplicationCookie(options =>\n{\n    options.Cookie.HttpOnly  = true;\n    options.Cookie.SecurePolicy = CookieSecurePolicy.Always;\n    options.Cookie.SameSite  = SameSiteMode.Strict;\n});" }
                                    }
                                });
                            }
                        }
                    }
                }
            }
            catch { }
        }

        return result;
    }
}

File: vscode extension/tests/sensitiveDataTest.js
const https = require("https");
const http  = require("http");
const { URL } = require("url");

const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

const PATTERNS = [
  { name: "Plain-text Password",      pattern: /["']?password["']?\s*[:=]\s*["'][^"']{3,}["']/i },
  { name: "Credit Card Number",       pattern: /\b(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|3[47][0-9]{13})\b/ },
  { name: "Stack Trace / File Path",  pattern: /at\s+\w+.*\.(?:cs|js):\d+|(?:\/home\/|\/var\/www\/|C:\\Users\\)/i },
  { name: "Database Connection String", pattern: /mongodb:\/\/|mysql:\/\/|postgres:\/\/|Server=.*;Database=/i },
  { name: "Internal IP Address",      pattern: /\b(?:192\.168\.|10\.|172\.(?:1[6-9]|2\d|3[01])\.)\d+\.\d+\b/ },
  { name: "JWT Token in Response",    pattern: /eyJ[a-zA-Z0-9_-]+\.eyJ[a-zA-Z0-9_-]+\.[a-zA-Z0-9_-]+/ },
];

const COOKIE_FLAGS = ["httponly", "secure"];

function fetchURL(url, timeout = 10000) {
  return new Promise((resolve) => {
    try {
      const parsed = new URL(url);
      const lib    = parsed.protocol === "https:" ? https : http;
      const req    = lib.get(url, { timeout, rejectUnauthorized: false }, (res) => {
        let data = "";
        res.on("data", (c) => { data += c; if (data.length > 500000) req.destroy(); });
        res.on("end", () => resolve({ status: res.statusCode, headers: res.headers, body: data }));
      });
      req.on("error",   () => resolve(null));
      req.on("timeout", () => { req.destroy(); resolve(null); });
    } catch { resolve(null); }
  });
}

async function testSensitiveData(target, siteMap) {
  const result  = { name: "Sensitive Data Exposure", passed: true, issues: [] };
  const baseUrl = target.replace(/\/$/, "");

  const pages = siteMap?.pages?.map(p => p.url) || [
    baseUrl, `${baseUrl}/api`, `${baseUrl}/api/users`, `${baseUrl}/login`
  ];

  for (const page of pages) {
    try {
      const res = await fetchURL(page);
      if (!res || res.status === 404) continue;
      await delay(100);

      const body = res.body;

      for (const p of PATTERNS) {
        const match = body.match(p.pattern);
        if (!match) continue;

        const raw    = match[0];
        const masked = raw.length > 10 ? raw.slice(0, 6) + "..." + raw.slice(-4) : "****";

        result.passed = false;
        result.issues.push({
          title:    `Sensitive Data Exposed: ${p.name}`,
          location: page,
          detail:   `Pattern matched in response: "${masked}"`,
          severity: p.name.includes("Password") || p.name.includes("Card") ? "CRITICAL" : "HIGH",
          risk:     `Exposing ${p.name.toLowerCase()} gives attackers direct access to credentials or system internals.`,
          fixes: [
            { label: "Never return sensitive fields in API responses",
              code:  "// ❌ VULNERABLE — returns everything including password hash\nres.json(user);\n\n// ✅ SAFE — use a DTO\nres.json({ id: user.id, name: user.name, email: user.email });" },
            { label: "Never expose stack traces in production",
              code:  "// In Express:\napp.use((err, req, res, next) => {\n  if (req.app.get('env') === 'development') {\n    return res.status(500).json({ error: err.stack });\n  }\n  res.status(500).json({ error: 'An error occurred' });\n});" }
          ]
        });
      }

      const setCookie = res.headers["set-cookie"];
      if (setCookie) {
        const cookies = Array.isArray(setCookie) ? setCookie : [setCookie];
        for (const cookie of cookies) {
          const cl = cookie.toLowerCase();
          for (const flag of COOKIE_FLAGS) {
            if (!cl.includes(flag)) {
              result.passed = false;
              result.issues.push({
                title:    `Cookie Missing "${flag}" Flag`,
                location: `Set-Cookie header at ${page}`,
                detail:   `Cookie set without "${flag}": ${cookie.split(";")[0]}`,
                severity: "MEDIUM",
                risk:     flag === "httponly"
                  ? "JavaScript can read the cookie — XSS can steal session tokens."
                  : "Cookie sent over plain HTTP — can be intercepted.",
                fixes: [
                  { label: "Set all security flags on cookies",
                    code:  "res.cookie('session', token, {\n  httpOnly: true,\n  secure:   true,\n  sameSite: 'strict'\n});" }
                ]
              });
            }
          }
        }
      }
    } catch { }
  }

  return result;
}

module.exports = testSensitiveData;
